### DISCUSSION TITLE
Why do Primary Constructors have a limitation that they must be a record class/struct to have a body that is just a semicolon?

### DISCUSSION THREAD

**Discussion by RoyAwesome** [dotnet/csharplang#6871] [General]
I was reading the Primary Constructor proposal, and this line jumped out at me:

> It is an error for a class_body or struct_body to consist of just a ; unless the corresponding class_declaration or struct_declaration has a record keyword and a parameter_list.

Why this restriction?  Some 

```cs
struct Foo(int x, int y);
```

seems useful.

**Comment by DavidArno** [comment]
See #6569. Primary constructors are still in design and hopefully will make it into the language soon(ish).

As @Richiban mentions below, under the current proposal for how primary constructors will work, `struct Foo(int x, int y);` won't produce anything particularly useful.

Due to C# being open source these days though, it's possible to play around with the current prototype, eg via the [Primary Constructors branch in SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQAmIA1AD6XV0AEAMRIkAFFjIw+ADyZ9xkggEoA3AFgAUBp61JAIXxiJ02fL5KNAbw18bfAAIBmOUYAafCwHMIMZQF8+ALzSauq29k6mAJruXj7+QQQhvkA===). This link highlights the difference in terms of code that you'll get with the following too structs:
```cs
struct Foo(int x, int y);

struct Bar(int x, int y)
{
    public int X {get;} = x;
    public int Y {get;} = y;
}
```

**Comment by Richiban** [comment]
@RoyAwesome 

> struct Foo(int x, int y);

As far as I'm aware, this would be a useless type because in non-record types the class parameters function as captures (that can be lifted into fields if required by the class members). That means that your type `struct Foo(int x, int y);` is functionally equivalent to:

```cs
struct Foo
{    
    public Foo(int x, int y) 
    {
    }
}
```

I can see an argument that it should be a warning rather than error, but it's still useless.

**Reply by HaloFour** [reply]
Correct, primary constructors on non-record types do not synthesize any members.  They are only captures which may be referenced by members within the type.

However, they could be used with a source generator which would emit members, which might be helpful for those who want different code generation than is provided by records.

**Reply by RoyAwesome** [reply]
@Richiban  Thanks for the reply.  That's a good reason I missed in my reading.  I was thinking they synthesized members, but given that it does not, that makes sense!

> However, they could be used with a source generator which would emit members, which might be helpful for those who want different code generation than is provided by records.

Perhaps partial classes with just a ; for a body could also not be an error?

**Comment by 333fred** [comment]
Thanks for bringing this up. We'll be discussing the restriction in LDM soon (likely next week, but no guarantees).

**Reply by 333fred** [reply]
Discussed in LDM: https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-01-18.md#semi-colon-bodies-for-type-declarations
